closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -380,30 +398,111 @@ JNIEXPORT jboolean JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_isOpenSSLFIP
     return OSSL_IS_FIPS;
 }
 
-/*
- * Class:     jdk_crypto_jniprovider_NativeCrypto
- * Method:    loadCrypto
- * Signature: (Z)J
- */
-JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
-  (JNIEnv *env, jclass thisObj, jboolean trace)
+/* Unload the crypto library. */
+static void
+unload_crypto_library(void *crypto_library)
 {
-    typedef const char* OSSL_version_t(int);
+#if defined(_WIN32)
+    FreeLibrary(crypto_library);
+#else /* defined(_WIN32) */
+    (void)dlclose(crypto_library);
+#endif /* defined(_WIN32) */
+}
 
-    /* Determine the version of OpenSSL. */
-    OSSL_version_t* OSSL_version;
-    const char * openssl_version;
-    jlong ossl_ver = 0;
+/* Find the symbol in the crypto library (return NULL if not found). */
+static void *
+find_crypto_symbol(void *crypto_library, const char *symname)
+{
+#if defined(_WIN32)
+    return GetProcAddress(crypto_library, symname);
+#else /* defined(_WIN32) */
+    return dlsym(crypto_library, symname);
+#endif /* defined(_WIN32) */
+}
 
-    /* Load OpenSSL Crypto library */
-    crypto_library = load_crypto_library(trace);
-    if (NULL == crypto_library) {
-        if (trace) {
-            fprintf(stderr, "Error loading OpenSSL: FAILED TO LOAD OPENSSL CRYPTO LIBRARY\n");
-            fflush(stderr);
+static void
+log_crypto_library_path(jboolean traceEnabled, void *crypto_library, const char *message)
+{
+    if (traceEnabled && (NULL != crypto_library)) {
+#if defined(_AIX)
+        int rc = 0;
+        char *buf = NULL;
+        struct ld_info *cur_info = NULL;
+        char *path = NULL;
+        char *member_name = NULL;
```

Comment:
All these can be scoped within the `if` statement on line 442; `path` and `member_name` can have type `const char *`:
```c
        if (0 == rc) {
            char *buf = buffer;
            for (;;) {
                struct ld_info *cur_info = (struct ld_info *)buf;
                const char *path = cur_info->ldinfo_filename;
                const char *member_name = path + strlen(cur_info->ldinfo_filename) + 1;
                if (('\0' != *member_name) && (NULL != strstr(path, "/libcrypto"))) {
                    fprintf(stdout, "%s: %s(%s)\n", message, path, member_name);
                    fflush(stdout);
                    break;
                }
                if (0 == cur_info->ldinfo_next) {
                    break;
                }
                buf += cur_info->ldinfo_next;
            }
```

Reply:
Done and changed to suggestion with `for` loop instead of `do-while` as well.